Issue related to window reload caused by selecting search checkboxes (#798)

juzen2003 opened the pull request:
- This pull request is related to issue #798 
- Local branch: search_reload_issue_798
- Previous window reload issue is caused by the difference between opus.selections and selections obtained from getSelectionsExtrasFromHash. Previously, in opus.selections, if selected value has white space, the white space will be stored as white space. But in selections, the white space is stored as %20 (URL encoding for white space). This difference will cause reload in opus.load function. Now all white space will be stored as %20 so opus.selections and selections are the same and window won't reload.

juzen2003 committed:
- In search.js (line 244), in search checkbox change event handler function,
  when a value is selected, before updating that value to opus.selections,
  replace white space to %20. This %20 will be used in URL (URL encoding).
- In search.js (line 257), use .push to add elements to the values array.
- In search.js (line 240-277), clean up code style, change '' to "".
- In hash.js (line 25), in updateHash function, replace white space with
  %20, this will be used in URL.
- In hash.js (line 99), in getSelectionsExtrasFromHash function, include
  %20 in selections. (Previously we will replace %20 with white space for
  comparison later one, but now we don't do that).
- In opus.js (line 562), in normalizedURLAPICall function, before updating
  URL, we replace white space with %20.

juzen2003 committed:
- In opus.js (line 562), search.js (line 244), hash.js (line 25), change
  replace("", "%20") to replace(/ /g, "%20"). This will make sure all
  white spaces are replaced with %20. If we do replace("", "%20"), only
  the first white space gets replaced with %20. And this will cause issue
  when the checkbox value has more than one white space.

dstopp commented on opus/application/static_media/js/search.js:
```diff
@@ -237,31 +237,31 @@ var o_search = {
-
-           if ($(this).is(':checked')) {
+           let id = $(this).attr("id").split("_")[0];
+           let value = $(this).attr("value").replace(/\+/g, "%2B");
```
i'm sort of confused why there is a two-step replace/regexp here.  If it cannot be combined, perhaps a comment indicating what's going on would be useful.

rfrenchseti reviewed (changes requested):
So in the currently deployed implementation (which is how things have worked for years), a "+" in the URL means "space" and a "%3B" in the URL means "+". These appear for instrument names like "Cassini ISS" and the Cassini ISS filters like "BL1+GRN", respectively. It's important for this to still be supported for legacy URLs, and I believe these changes break that backwards compatibility.

The whole reason that "+" is replaced with "%3B" is to avoid it conflicting with the "+" that meant "space". It's an annoying hacked workaround that is now getting us in trouble. Honestly I don't see a clean fix to this. We probably have to take advantage of the fact that COISSfilter is the only field that has ever had "+" in the values, and it doesn't ever have spaces.

Here's what I suggest:

- Modify normalizeurl so that when a URL comes in, any existing "+" is replaced with "%20" for any slug value EXCEPT COISSfilter (or its old-slug equivalent). For COISSfilter, leave "+" as-is (this will be needed for the future to avoid continually rewriting this field; see below)
- Modify normalizeurl so that when a URL comes in, any existing "%3B" is replaced with "+" (in that order, of course) (this is safe for both old and new URLs)
- Going forward, leave "+" as-is in the URL - don't encode it as "%3B" anymore
- Going forward, encode "space" as "%20"
- Since replacing "space" with "%20" is really just an escaping of "dangerous" characters in the URL, we should do that with all dangerous characters (like "&"). See issue #719 for details on this. I think solving #719 will automatically solve the "space" issue here as well.
- Make sure all this is carefully commented so we understand it in the future

See JavaScript's "encodeURI" and "decodeURI" for how to do this cleanly.

juzen2003 committed:
- In search.js (line 241-280), fix the indentation.
- In search.js (line 243-246), add comments that we will convert space to
  "%20" and "+" to "%2B" when selecting values of checkboxes.
- In opus.js (line 564-569), modify normalizedURLAPICall so that new url
  will always have "%20" as space and "%2B" as "+".
- In hash.js (lin 25), remove redundant conversion of space to "%20" in
  opus.selections.

juzen2003 committed:
- In opus.js (line 159), remove unwanted console.log in load function.

juzen2003 committed:
- In search.js (line 243), update on change event handler function. Now
  we don't convert "+" to "%2B", we will store the unencoded values in
  opus.selections.
- In opus.js (line 561-576). modify normalizedURLAPICall function. Before
  updating new URL, we will encode it first.
- In hash.js (line 49-109), create "encodeSlugValue", "encodeHashArray",
  and "decodeHashArray" functions to perform hash encoding/decoding. More
  details in comments of each function.
- In hash.js (line 149, 162-167), modify getSelectionsExtrasFromHash to
  decode hash before building selections from it.
- In hash.js (line 202, 242), modify initFromHash to decode hash before
  building opus.selections from it.
- In hash.js (line 24-25, 30, 32, 35), modify updateHash to encode values
  in opus.selections before building hash from it.

juzen2003 committed:
- In hash.js (around line 239 area), remove unwanted comments.

juzen2003 committed:
- In hash.js (line 242-249), modify initFromHash, update the logic to make
  "," in slug value gets encoded in "%2C" in URL, and store as decoded ","
  in opus internal variables opus.selections.
- In hash.js (line 162-164), update comments in getSelectionsExtrasFromHash.

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -18,20 +18,21 @@ var o_hash = {
 
         for (let key in opus.extras) {
+            let encodedExtraValues = o_hash.encodeSlugValue(opus.extras[key]);
             try {
```
Since encodeSlugValue always returns an array now, we probably don't need this try anymore.

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -45,6 +46,67 @@ var o_hash = {
         return hash.join("&");
     },
 
+    encodeSlugValue: function(slugValueArray) {
```
This probably should be called encodeSlugValues since it can encode more than one value

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -45,6 +46,67 @@ var o_hash = {
+    encodeSlugValue: function(slugValueArray) {
+        /**
+         * Take in a slug value array (like opus.selections, each element
+         * will be a value for the slug) and encode all values in the
```
will be a "list of values" for the slug

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -45,6 +46,67 @@ var o_hash = {
+         * slug values in the hash are all encoded before updating the URL.
+         */
+        let slugValue = [];
+        $.each(slugValueArray, function(index, val) {
```
Given ES6, Debby and I talked about the use of $.each. It's really not needed for such a simple case and has excess overhead. It might be more readable to use the new "for...of" construct:
`for (const val of slugValueArray) {`

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -45,6 +46,67 @@ var o_hash = {
+        /**
+         * Take in a hash array (each element will be "slug=value") and
+         * encode the "value" of each element. Return a hash array that
+         * is consist of "slug=encodedValue". This function will be called
```
"that consists of"

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -45,6 +46,67 @@ var o_hash = {
+         * encode the "value" of each element. Return a hash array that
+         * is consist of "slug=encodedValue". This function will be called
+         * in opus.normalizedURLAPICall to make sure slug values from
+         * new URL are decoded.
```
Do you mean "are ENcoded"?

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -45,6 +46,67 @@ var o_hash = {
+         * new URL are decoded.
+         */
+        let hash = [];
+        $.each(hashArray, function(index, pair) {
```
Same use of `for...of` here

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -45,6 +46,67 @@ var o_hash = {
+        let hash = [];
+        $.each(hashArray, function(index, pair) {
+            let slug = pair.split('=')[0];
+            let value = pair.split('=')[1];
```
It's possible for a search string to have an "=" in it. For example, try searching on "a=b". Because the URL will say "searchslug=a=b", the split will break this into 3 parts, and we will only take "a" as the value. You can easily try this by searching a string field on "a=b" and notice that the URL will only say "a". It would be better to take value = ALL of the string after the first "=".

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -45,6 +46,67 @@ var o_hash = {
+        $.each(hashArray, function(index, pair) {
+            let slug = pair.split('=')[0];
+            let value = pair.split('=')[1];
+            value = encodeURIComponent(value);
```
So we have this legacy code everywhere that assumes values (and qtypes) are *lists*, not just single values. This is because the old version of OPUS allowed multiple search values for a given field, e.g. "observationduration1=1,2,3" and "observationduration2=4,5,6". We might actually want to allow this again at some point, so we might as well support it everywhere. It seems like the better thing to do here, rather than just do a single encodeURIComponent, is to split the value at ",", then call encodeSlugValue, and then join the results back with "," again. This also means the encode is only done in one place, so we don't have duplicate code that can diverge over time.

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -45,6 +46,67 @@ var o_hash = {
+        /**
+         * Take in a hash array (each element will be "slug=value") and
+         * decode the "value" of each element. Return a hash array that
+         * is consist of "slug=decodedValue". This function will be called
```
"that consists of"

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -45,6 +46,67 @@ var o_hash = {
+         * in initFromHash to make sure slug values in opus.selections are
+         * decoded.
+         */
+        // console.log(`=== hash from decodeHashArray ===`);
```
Remove comment

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -45,6 +46,67 @@ var o_hash = {
+         */
+        // console.log(`=== hash from decodeHashArray ===`);
+        let hash = [];
+        $.each(hashArray, function(index, pair) {
```
`for...of`

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -45,6 +46,67 @@ var o_hash = {
+        let hash = [];
+        $.each(hashArray, function(index, pair) {
+            let slug = pair.split('=')[0];
+            let value = pair.split('=')[1];
```
Same deal with split taking all the remaining parts of the string

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -96,7 +159,14 @@ var o_hash = {
+                    // If value contains ", ", it's one value from string input. This will make sure "," in
+                    // slug value gets encoded in "%2C" in URL, and store as decoded "," in opus internal variables
+                    // selections.
+                    if (value.includes(", ")) {
```
Try to look for the widget to see if it's a string instead of using includes.

juzen2003 replied:
Using $(`input[name="${slug}"]`).hasClass("STRING") to look for widget with string input will work, but when opus is reload, the page will keep reloading if there is an input with "," in string field. The reason is because  in opusInitialization, the selector will return false because the widget is not rendered yet at this point. And at the same time, both selections (from getSelectionsExtrasFromHash) and opus.selections (from initFromHash) are build. After opusInitialization, load is called and selections will get updated correctly (from getSelectionsExtrasFromHash) and it will be different from opus.selections, which will cause page reload again. (This will happen over and over).

So for now, we didn't update this part and when there a "," in string input, the page will reload one time. And if there is no "," in the string field, page will not reload and search will be performed directly.

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -169,7 +239,14 @@ var o_hash = {
+                    // If value contains ", ", it's one value from string input. This will make sure "," in
+                    // slug value gets encoded in "%2C" in URL, and store as decoded "," in opus internal variables
+                    // opus.selections.
+                    if (value.includes(", ")) {
```
Also look for the widget here

rfrenchseti commented on opus/application/static_media/js/opus.js:
```diff
@@ -558,8 +558,23 @@ var opus = {
+            // to get newURLHash array.
+            let newSlugArr = normalizeURLData.new_slugs;
+            let newURLHash = [];
+            for (let idx in newSlugArr) {
```
New ES6 uses "for...of"

juzen2003 committed:
- In opus.js (line 567), replace the for loop with ES6 for loop (
  for (let x of y) {}).
- In hash.js (line 46), rename "encodeSlugValue" to "encodeSlugValues".
- In hash.js (line 30-31), remove the try and catch for opus.extras because
  encodeSlugValues will always return an array.
- In hash.js (line 49,68,70,112), clean up comments.
- In hash.js (line 90-106), create a help function decodeSlugValues to be
  called in decodeHashArray. This function will help to decode each value
  in the value list of each slug (please see function description).
- In hash.js (line 73-85, 119-129), update the logic in encodeHashArray
  and decodeHashArray. Use ES6 for loop. Update the way to get the slug
  value by getting the first idx of "=" sign and take all of string after
  the first "=" as the value (or a list of values) for the slug. Also
  update to encode/decode each value in the value list of each slug.
- In hash.js (line 178-180, 230-232), in getSelectionsExtrasFromHash and
  initFromHash update the method to get the slug value by getting the
  first idx of "=" sign and take all of string after the first "=" as the
  value (or a list of values) for the slug.
- In hash.js (line 187-193, 266-273), update the way we build selections
  and opus.selections, string input value with "," will be split and treated
  as multiple values in these variables. So every time we input a string
  value with "," it will reload.

rfrenchseti commented on opus/application/static_media/js/hash.js:
```diff
@@ -45,6 +43,94 @@ var o_hash = {
+         * slug values in the hash are all encoded before updating the URL.
+         */
+        let slugValue = [];
+        for (let val of slugValueArray) {
```
For this and all other "for (let..." statements we should use "for (const..." instead assuming the value isn't going to be modified in the loop.
